### DISCUSSION TITLE
Its possible to have a C++ static_assert equivalent on dotnet?

### DISCUSSION THREAD

**Discussion by gustavopsantos** [dotnet/runtime#80826] [Ideas]
I would like to know, from compiler perspective, its possible to have assertions that runs before code gets compiled?
It can be very useful to do some safe guards.

**Comment by KalleOlaviNiemitalo** [comment]
In C#, some compile-time assertions can be done with hacks like this ([SharpLab](https://sharplab.io/#v2:CYLg1APgAgTABAYTgbwLACg5blAzHAYwHsA7AZwBc4BLEqgSTgF44A2ATgG4Nsd9aqAWQCmAWwBGwgE4o4Ac2EVOcAL4Y16DFACMrHPACCKHtgD0puABEicEkSoEAhgBtncR1SkBXEnArVRYQA6EywdPSgAFjgAZQoPagIDMjJpf1IyAAokWkpHEgJhAEpjTF4zC2TUqSoKAAsPRCDGajI4SJgQsvK4AH1mW2EAdzhxAE8KYQBtBGbmFg64AH44bTgQOABabQBdbm7y0PLzOCq0vwbauuE/MYAHG6IAMyaRCWkaNudSOS6erAAbo4ZKIBrl4gVgm9JFJ9v9YvF/EkUmkACr3YQAHm+JDkAD5MlJhC9REU4dgNLwjuEcNE4glkdUKOiHpjUQSiS9UXAgc4vMUUBoNEA==)):

```C#
class C {
    public const int I = 69;
    public int Member { get; }
}

static class A {
    // Do not call at run time.
    static void StaticAssertions(C instance) {
        // Assert that C.I is 42.
        _ = new byte[C.I == 42 ? 1 : -1];
        
        // Assert that the type of C.Member is long.
        var m = instance.Member;
        StaticAssertType<long>(ref m);
    }
    
    static void StaticAssertType<T>(ref T value) {}
}
```

If the assertions fail, the compiler errors do not outright say so:

```
error CS0248: Cannot create an array with a negative size
error CS1503: Argument 1: cannot convert from 'ref int' to 'ref long'
```

A unit test, or perhaps a Roslyn analyzer, could give better diagnostics.
